File: 0001-0050/0005/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        string ans = "";
        for (int i = 0; i < s.length(); i++) {
            // 奇数长度的回文串
            int l = i - 1, r = i + 1;
            while (l >= 0 && r < s.length() && s[l] == s[r]) l--, r++;
            if (ans.length() < r - l - 1) ans = s.substr(l + 1, r - l - 1);
            
            // 偶数长度的回文串
            l = i, r = i + 1;
            while (l >= 0 && r < s.length() && s[l] == s[r]) l--, r++;
            if (ans.length() < r - l - 1) ans = s.substr(l + 1, r - l - 1);
        }
        return ans;
    }
};

struct Example {
    string s;
    string ans;
};

int main() {
    vector<Example> examples = {
        {"babad", "bab"},
        {"cbbd", "bb"},
    };
    
    Solution *solve = new Solution();
    for (int i = 0; i < examples.size(); i++) {
        string ans = solve->longestPalindrome(examples[i].s);
        if (ans == examples[i].ans) {
            cout << "PASS: CASE " << i << endl;
        } else {
            cout << "FAIL: CASE " << i << endl;
        }
    }
    
    return 0;
}

File: 0001-0050/0005/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        ans = ""
        for i in range(len(s)):
            # 奇数长度的回文串
            l, r = i - 1, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            if len(ans) < r - l - 1:
                ans = s[l + 1:r]

            # 偶数长度的回文串
            l, r = i, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            if len(ans) < r - l - 1:
                ans = s[l + 1:r]
        return ans


class Example:
    def __init__(self, s, ans):
        self.s = s
        self.ans = ans


def main():
    examples = [
        Example("babad", "bab"),
        Example("cbbd", "bb"),
    ]

    solve = Solution()
    for i, example in enumerate(examples):
        ans = solve.longestPalindrome(example.s)
        if ans == example.ans:
            print(f"PASS: CASE {i}")
        else:
            print(f"FAIL: CASE {i}")


if __name__ == "__main__":
    main()
